src/deploy/functions/backend.ts

```diff
@@ -0,0 +1,356 @@
+
+  // What we're actually planning to invoke with this schedule
+  targetService: FunctionSpec;
+}
+
+export interface HttpsTrigger {
+  httpsOnly: boolean;
+}
+
+export type EventFilterKey = "resource";
+export interface EventTrigger {
+  eventType: string;
+
+  // In v1 API, only "resource" is set.
+  // In v2 API, "resource" will be hoisted up as
+  // a top-level "pubsubTopic" resource for the
+  // google.cloud.pubsub.topic.v1.messagePublished event.
+  eventFilters: Record<EventFilterKey | string, string>;
+
+  retry: boolean;
+
+  // API v2 only. Defaults to "global".
+  region?: string;
+  serviceAccountEmail?: string;
+}
+
+export function isEventTrigger(trigger: HttpsTrigger | EventTrigger): trigger is EventTrigger {
+  return "eventType" in trigger;
+}
+
+export type VpcEgressSettings = "PRIVATE_RANGES_ONLY" | "ALL_TRAFFIC";
+export type IngressSettings = "ALLOW_ALL" | "ALLOW_INTERNAL_ONLY" | "ALLOW_INTERNAL_AND_GCLB";
+export type MemoryOptions = 128 | 256 | 512 | 1024 | 2048 | 4096;
+export type Runtime = "nodejs6" | "nodejs8" | "nodejs10" | "nodejs12" | "nodejs14";
+
+export interface FunctionSpec {
+  apiVersion: 1 | 2;
+  id: string;
+  region: string;
+  project: string;
+  entryPoint: string;
+  trigger: HttpsTrigger | EventTrigger;
+  runtime: Runtime;
+
+  labels?: Record<string, string>;
+  environmentVariables?: Record<string, string>;
+  availableMemoryMb?: MemoryOptions;
+  timeoutSeconds?: number;
+  maxInstances?: number;
+  minInstances?: number;
+  vpcConnector?: string;
+  vpcConnectorEgressSettings?: VpcEgressSettings;
+  ingressSettings?: IngressSettings;
+  serviceAccount?: "default" | string;
+}
+
+// Note(inlined): I'd like to put topics in here explicitly, but I'd first have to
+// do some refactoring so that we know the GAE region by the time we create the Backend
+// spec.
+export interface Backend {
+  requiredAPIs: Record<string, string>; // friendly-name -> API name
+  cloudFunctions: FunctionSpec[];
+  schedules: ScheduleSpec[];
+  topics: PubSubSpec[];
+}
+
+export const EMPTY: Backend = Object.freeze({
+  requiredAPIs: {},
+  cloudFunctions: [],
+  schedules: [],
+  topics: [],
+});
+
+export function isEmptyBackend(backend: Backend) {
+  return (
+    Object.keys(backend.requiredAPIs).length == 0 &&
+    backend.cloudFunctions.length === 0 &&
+    backend.schedules.length === 0 &&
+    backend.topics.length === 0
+  );
+}
+
+export type RuntimeConfigValues = Record<string, any>;
+
+export function functionName(cloudFunction: FunctionSpec) {
+  return `projects/${cloudFunction.project}/locations/${cloudFunction.region}/functions/${cloudFunction.id}`;
+}
+
+// Curried function that's useful in filters. Compares fields in decreasing entropy order
+// to short circuit early (not like there's much point in optimizing JS...)
+export const sameFunctionName = (func: FunctionSpec) => (test: FunctionSpec) => {
+  return func.id === test.id && func.region === test.region && func.project == test.project;
+};
```

Comment:
This feels a little overclever maybe? This just take two function specs instead of being a curried function.